sound/soc/sof/ipc4-pcm.c

```diff
@@ -495,17 +496,44 @@ static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
 				use_chain_dma = true;
 		}
 	}
+
+	available_fmt = &ipc4_copier->available_fmt;
+	fe_rate = params_rate(params);
+
 	/*
-	 * Always set BE format to 32-bits for both playback and capture, except if chained
-	 * DMA is in use. In chained DMA case the back-end and front-end format must match.
+	 * Limit sampling rate if copier has a format that matches
+	 * the FE rate.
 	 */
 	if (!use_chain_dma) {
```

Comment:
so now if (use_chain_dma) we do nothing but the original code set rate->min and rate->max.  If we don't need to set rate
mix and max, we can do it as
````c
if (use_chain_dma)
     return 0
````

Reply:
@RanderWang I was wondering about this but in theory we could use chain-dma for SSP as well, and in that case, we may need to run the DAI-type specific logic later in the function. But yeah, this code could be more reabable, let me split this segment to a separate subfunction in next version.

Reply:
We cannot use chain DMA for SSP. It uses gpdma

Reply:
@ranj063, #4158 gives a glimpse into the future ;)
But the ChainDMA FW implementation as it is does not rule out use of GPDMA on the peripheral side, with some effort it can be made working.

Reply:
Having said that, for now a simple return 0 in case of chain_dma would make the code nicer. If there will be a need for a change, it can be done when it is due.

Reply:
@ranj063 @ujfalusi As it's more a FW side development whether chain_dma can work with SSP, I'd rather keep the kernel code neutral on this.